cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -236,35 +225,33 @@ private boolean checkJobConstraints(DatasetContext context) throws Exception {
           emptyScan = false;
           checkAndUpdateJob(jobQueue, job);
         }
-      } finally {
-        jobQueueIter.close();
       }
       return emptyScan;
     }
 
     private void checkAndUpdateJob(JobQueueDataset jobQueue, Job job) {
-      switch (job.getState()) {
-        case PENDING_TRIGGER:
-          if (isTriggerSatisfied(job)) {
-            jobQueue.transitState(job, Job.State.PENDING_CONSTRAINT);
-          }
-        // fallthru to check for constraint satisfaction after trigger satisfaction
-        case PENDING_CONSTRAINT:
-        case PENDING_LAUNCH:
-          if (constraintsSatisfied(job)) {
-            jobQueue.transitState(job, Job.State.PENDING_LAUNCH);
-            readyJobs.add(job);
-          }
-          break;
+      if (job.getState() == Job.State.PENDING_TRIGGER) {
+        if (!isTriggerSatisfied(job)) {
+          return;
+        }
+        job = jobQueue.transitState(job, Job.State.PENDING_CONSTRAINT);
+      }
+      // check for constraint satisfaction after trigger satisfaction
+      if (!constraintsSatisfied(job)) {
```

Comment:
If the system crashes after marking a Job as `PENDING_LAUNCH`, then when it comes back up, it needs to check the constraints again, since any time range constraint may no longer be satisfied.